crypto/evp/e_eea3.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2021 Ant Group. All Rights Reserved.
+ * Copyright 2018 BaishanCloud. All Rights Reserved.
+ */
+
+#include "internal/cryptlib.h"
+
+#ifndef OPENSSL_NO_ZUC
+
+# include <openssl/evp.h>
+# include <openssl/objects.h>
+
+# include "crypto/zuc.h"
+# include "crypto/evp.h"
+
+typedef struct {
+    ZUC_KEY zk;                 /* working key */
+} EVP_EEA3_KEY;
+
+# define data(ctx) ((EVP_EEA3_KEY *)EVP_CIPHER_CTX_get_cipher_data(ctx))
+
+static int eea3_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
+                        const unsigned char *iv, int enc);
+static int eea3_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                      const unsigned char *in, size_t inl);
+static int eea3_cleanup(EVP_CIPHER_CTX *ctx);
+
+static const EVP_CIPHER zuc_128_eea3_cipher = {
+    NID_zuc_128_eea3,
+    1, EVP_ZUC_KEY_SIZE, 5,
```

Comment:
The length of iv is 5 bytes? should it be 16 bytes?

Reply:
EEA3 calls ZUC to generate the keystream, which requires a 16 bytes IV value. But the IV is not fully random, the IV is formed by three parameters: count, bearer and direction. OpenSSL's EVP framework doesn't support passing those parameters into a cipher, so I use the IV to hold those three parameters and then build the real IV inside EEA3...and then call ZUC to generate the keystream. The parameter count, bearer and direction require 38 bits to store, so 5 bytes (40 bits) are allocated here to hold those values.

Reply:
Check `eea3_init_key()` function for details.